### ISSUE TITLE
Plan to Open Source

### ISSUE THREAD

**Issue by yangelaboy** [NVlabs/Jet-Nemotron#1]
I am interested in your great work, so I want to known any plan to open source.

**Comment by t1101675** [comment]
Thanks for your attention to our work! Our code and data are still under legal review. We will open-source them as soon as we get permission.

**Comment by blart4525-design** [comment]
> Thanks for your attention to our work! Our code and data are still under legal review. We will open-source them as soon as we get permission.

you are doing great work

**Comment by isayoften** [comment]
@t1101675 Great! Will you open source code too?

**Comment by t1101675** [comment]
> [@t1101675](https://github.com/t1101675) I’m not sure how Linear Attention Block Selection and Hardware-Aware Architecture Search were trained. Was it the same procedure as Full Attention Placement and Elimination (freezing MLPs with MSE loss), or did you use full fine-tuning?

We freeze MLPs and use MSE loss when doing Linear Attention Block Selection and Hardware-Aware Architecture Search. After the final architecture is determined, we use full fine-tuning to train the model.

**Comment by t1101675** [comment]
A1: The dynamic convolutional block generates the convolution kernel from the value feature by an MLP with one hidden layer. Then the kernel weights in torch.nn.Conv1d is replaced by the generated weights to perform convolution.
A2: It is an MLP with one hidden layer.
A3: Time mixing is a general term for the operations on the time (sequence length) axis, such as full-attention, gated deltanet, mamba2, etc.
A4: We maintain the state sizes of JetBlock and conduct a grid search on these dimensions. For each dimension choice, we randomly initialize the attention blocks of the model and train these layers only with the MLP layers frozen.
A5: We only use 2 full-attention blocks, and the number of heads are inherited from Qwen2.5. The search for the number of heads only happens in the efficient attention (JetBlock) layers.

Adding convolution layers to linear attention blocks is proven to be beneficial for modeling local information and final performance without sacrificing inference speed [in](https://arxiv.org/abs/2405.21060) [many](https://arxiv.org/abs/2205.14756) [recent](https://papers.ssrn.com/sol3/papers.cfm?abstract_id=5240330) [works](https://arxiv.org/abs/2412.06464).

**Comment by indramal** [comment]
> A1: The dynamic convolutional block generates the convolution kernel from the value feature by an MLP with one hidden layer. Then the kernel weights in torch.nn.Conv1d is replaced by the generated weights to perform convolution. A2: It is an MLP with one hidden layer. A3: Time mixing is a general term for the operations on the time (sequence length) axis, such as full-attention, gated deltanet, mamba2, etc. A4: We maintain the state sizes of JetBlock and conduct a grid search on these dimensions. For each dimension choice, we randomly initialize the attention blocks of the model and train these layers only with the MLP layers frozen. A5: We only use 2 full-attention blocks, and the number of heads are inherited from Qwen2.5. The search for the number of heads only happens in the efficient attention (JetBlock) layers.
> 
> Adding convolution layers to linear attention blocks is proven to be beneficial for modeling local information and final performance without sacrificing inference speed [in](https://arxiv.org/abs/2405.21060) [many](https://arxiv.org/abs/2205.14756) [recent](https://papers.ssrn.com/sol3/papers.cfm?abstract_id=5240330) [works](https://arxiv.org/abs/2412.06464).

Thank you very much for informations

**Comment by EwoutH** [comment]
> Our code and data are still under legal review. We will open-source them as soon as we get permission.

Any update to share?

**Comment by liudd-1** [comment]
I have a question：two sliding window attention layers were added to the model in the end, what scheme was used to decide where to place these two layers?
